File: 0677-map-sum-pairs/0677-map-sum-pairs.py
from typing import Dict

class TrieNode:
    def __init__(self):
        self.children: Dict[str, 'TrieNode'] = {}
        self.is_end_of_word = False
        self.sum = 0


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str, val: int) -> None:
        current = self.root
        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()
            current = current.children[c]
            current.sum += val
        current.is_end_of_word = True

    def starts_with(self, prefix: str) -> int:
        current = self.root
        for c in prefix:
            if c not in current.children:
                return 0
            current = current.children[c]
        return current.sum


class MapSum:
    def __init__(self):
        self.map: Dict[str, int] = {}
        self.trie = Trie()

    def insert(self, key: str, val: int) -> None:
        curr_val = self.map.get(key, 0)
        self.map[key] = val
        self.trie.insert(key, val - curr_val)

    def sum(self, prefix: str) -> int:
        return self.trie.starts_with(prefix)

File: 0677-map-sum-pairs/0677-map-sum-pairs.java
class MapSum {

    private Map<String, Integer> map;
    private Trie trie;

    public MapSum() {
        map = new HashMap<>();
        trie = new Trie();    
    }
    
    public void insert(String key, int val) {
        int currVal = map.getOrDefault(key, 0);
        map.put(key, val);
        trie.insert(key, val - currVal);
    }
    
    public int sum(String prefix) {
        return trie.startsWith(prefix);
    }
}

class TrieNode {
    TrieNode[] children;
    boolean isEndOfWord;
    int sum;

    public TrieNode() {
        children = new TrieNode[26];
        isEndOfWord = false;
        sum = 0;
    }
}
class Trie {

    private TrieNode root;

    public Trie() {
        root = new TrieNode();    
    }
    
    public void insert(String word, int val) {
        TrieNode current = root;

        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (current.children[idx] == null) {
                current.children[idx] = new TrieNode();
            }

            current.children[idx].sum += val;

            current = current.children[idx];
        }
        current.isEndOfWord = true;
    }
    
    public boolean search(String word) {
        TrieNode current = root;

        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (current.children[idx] == null) {
                return false;
            }
            current = current.children[idx];
        }
        return current.isEndOfWord;
    }
    
    public int startsWith(String prefix) {
        TrieNode current = root;

        for (char c : prefix.toCharArray()) {
            int idx = c - 'a';
            if (current.children[idx] == null) {
                return 0;
            }
            current = current.children[idx];
        }
        return current.sum;
    }
}
